timestamp-tests-shared/src/main/java/com/palantir/timestamp/RateLimitedTimestampServiceTest.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright 2015 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.timestamp;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.Test;
+
+import com.palantir.common.concurrent.PTExecutors;
+
+public class RateLimitedTimestampServiceTest {
+    @Test
+    public void testRateLimiting() throws Exception {
+        final long MIN_REQUEST_MILLIS = 100L;
+        final long TEST_DURATION_MILLIS = 2000L;
+        final int NUM_THREADS = 3;
+
+        final StatsTrackingTimestampService rawTs = new StatsTrackingTimestampService(new InMemoryTimestampService());
+        final RateLimitedTimestampService cachedTs = new RateLimitedTimestampService(rawTs, MIN_REQUEST_MILLIS);
+        final AtomicLong timestampsGenerated = new AtomicLong(0);
+        final long startMillis = System.currentTimeMillis();
+
+        ExecutorService executor = PTExecutors.newCachedThreadPool();
```

Comment:
There's a Mockito class called DeterministicScheduler which is a deterministic `ExecutorService`, can we use that instead of the timing here? (we should actually convert most of our tests to use this, as the timing stuff is super flakey on travis)